feat: register_custom_background_input_handler() API

bolinfest opened the pull request:
This introduces a new API on `ESCInputsManager` that makes it
possible to inject custom logic for processing input events.
https://github.com/bolinfest/escoria-demo-game/commit/ae786cfd17e080f99a0194a125ce055ccc433531
is an example of how this could be used to add support for keyboard
shortcuts to select a specific verb.

bolinfest commented:
I could also see this as:

```
        var is_default_state = not escoria.current_state == escoria.GAME_STATE.DEFAULT
	if escoria.inputs_manager.handle_background_input(event, is_default_state):
		return
	if not is_default_state:
		return
```

where the caller can decide whether they want to respect the game state or not.

bolinfest commented:
OK, with the two-arg version, sample code that uses the API becomes:

```
func _background_input_handler(event: InputEvent, is_default_state: bool) -> bool:
	if not is_default_state:
		return false
```

dploeger commented:
This is a great idea. I'm just wondering, if we should rename the handler, because in your demo you don't really care about the background. It's a shortcut handler. Maybe "custom_input_handler" or something?

dploeger reviewed (changes requested):
Some smaller remarks and please see my latest comment.

bolinfest commented:
Hmm, when the handler returns`true`, we should probably call `set_input_as_handled ( )` on it, right? Any reason why that should be configurable?

BHSDuncan commented:
Might be a good idea to make mention of this in the docs, esp in the "how to" section.

bolinfest commented:
@dploeger I did my best to try to follow existing documentation standards. Feel free to edit yourself and commit or kick back to me for follow-up.

@BHSDuncan adding to docs is a good call: I can do that once this lands. That would be in https://github.com/godot-escoria/escoria-docs, I assume?

bolinfest commented:
Also, do you prefer I squash these three commits into one? Different projects have different preferences, in my experience.

dploeger commented:
Thanks! Comments look great (we actually didn't use a "returns" section, but I like it more like this).

For the docs: We would have two locations here. For one, we could use the doc comment to document that, because that is generated into the API documentation and we could write an article in the advanced section about input handling. Actually, there's an open issue godot-escoria/escoria-issues#33 about that, that is assigned to @StraToN .

dploeger reviewed (approved):
Great contribution! Thanks a lot! 🎁

dploeger commented:
> Also, do you prefer I squash these three commits into one? Different projects have different preferences, in my experience.

Oh, yes. One commit at the end of the PR would be great for history clean-ness. You can rebase or we can squash it when merging.

BHSDuncan commented on addons/escoria-core/game/inputs_manager.gd:
```diff
@@ -31,7 +31,19 @@ var hover_stack: Array = []
 var hotspot_focused: String = ""
 
-var custom_background_input_handler = null
+# Function that can be used to intercept and process input events.
```
```suggestion
# Function reference that can be used to intercept and process input events.
```

Just to make it painfully clear that `var custom_input_handler` isn't actually a function implementation (call me pedantic).

bolinfest replied:
I added "reference" here and in one other docstring.

bolinfest commented on addons/escoria-core/game/inputs_manager.gd:
```diff
@@ -102,20 +114,30 @@ func register_background(background: ESCBackground):
-func handle_background_input(event: InputEvent, is_default_state: bool) -> bool:
-	if custom_background_input_handler:
-		return custom_background_input_handler.call_func(event, is_default_state)
+# If a callback was specified via `register_custom_input_handler()`,
```
@BHSDuncan I'm not sure exactly what you mean, but I believe the challenge is we are documenting a parameter (whose type is function) as well as the contract of the function, and I didn't find an existing instance of how that is done handy, so I had to make something up. Since only `callback` is the parameter, is the only thing explicitly listed under `#### Parameters`

bolinfest replied:
(I was looking in the other view and I think the line numbers shifted since the comment was made...)

bolinfest replied:
OK, I believe this is fixed now!

Perhaps this raises the question of whether the call to `get_tree().set_input_as_handled()` should be moved from `esc_background.gd` to `inputs_manager.gd`?

BHSDuncan replied:
Sorry, I probably wasn't clear enough (wasn't sure how to select multiple lines).

I tried to highlight the first docstring for `try_customer_input_handler` to indicate that entire comment block: I was hoping you could format it more like the other functions, i.e.

```
<description>

<params>

<returns>
```

But good improvement anyway!

bolinfest replied:
Ugh, I hate Git so much...I forgot to add to the index before that last update...

bolinfest replied:
OK, the requested bits are there now.

bolinfest commented:
@BHSDuncan I’m happy to add that, though do you have a preferred application? And should it become part of the default 9verb or simple mouse? Maybe behind Options to enable keyboard shortcuts?”

BHSDuncan commented:
> @BHSDuncan I’m happy to add that, though do you have a preferred application? And should it become part of the default 9verb or simple mouse? Maybe behind Options to enable keyboard shortcuts?

You raise a good point  that might warrant a larger discussion: Personally, I don't think it should be part of the stock/default UI addons, but if that's the case, then we'd probably need a fork/version of one of the UIs to make into its own addon and then have it available in the demo game. At least, that would be the naïve approach; perhaps there's a better/smarter way.

@dploeger @StraToN ^^

bolinfest commented:
Thinking about the demo game some more, note that my sample code required updating project.godot to add entries to the input map, which may or may not be acceptable for the demo game?

StraToN commented:
> Thinking about the demo game some more, note that my sample code required updating project.godot to add entries to the input map, which may or may not be acceptable for the demo game?

@dploeger This introduces a new requirement we didn't expect: exclude some code so that it is not committed in escoria-core. 
I fear having these new elements in project.godot file could be problematic for users.

dploeger commented:
I'd recommend creating a new UI plugin for that as a demonstration and yes I'd really appreciate this to test the function.

That plugin could add the input maps and remove them if unloaded, so we wouldn't need to exclude code from core or any other plugin for it.

@bolinfest If you'd add them to your PR I'll add a new repo for it and the workflow to distribute the plugin to that repo and all.

Thanks a lot! I hope it's not too much to ask. ❤️

dploeger commented:
(oh and adding the docs afterwards is perfectly fine! It's a separate repo anyway)

bolinfest commented:
@dploeger do you want the new UI plug-in as part of this PR *and* this commit, or separated in some way?

bolinfest commented:
Though personally, I certainly wouldn’t mind just landing this commit to establish some progress on this issue :P

dploeger commented:
I'd like it as part of the PR, yes, but not necessarily in one commit. Without it, we can't validate that the feature works at all and test it later. (bringing some tdd into this 😆 )

bolinfest commented:
@dploeger OK, will figure it out. Indeed, `InputMap` has an `action_add_event()` API, but it isn't that clear how to use it, though this thread has a concrete example: https://godotengine.org/qa/30476/how-to-add-an-inputevent-to-inputmap-with-code

i.e., something like:

```
var ev = InputEventKey.new()
ev.scancode = KEY_A
InputMap.add_action("ui_a")
InputMap.action_add_event("ui_a", ev)
```

I should make it clear in the code that I write that although it's *possible* to create an `InputMap` this way, the recommended way is through the Godot project manager (unless you want to give the player the ability to customize shortcuts, I guess?).

bolinfest commented:
@dploeger OK, I just added a new commit that introduces `addons/escoria-ui-keyboard-9verbs`. I only wired up one keyboard shortcut for now because I wanted to get feedback on this direction before proceeding.

bolinfest commented:
Also, I don't know if this is expected, but `_enter_tree()` appears to be called twice, which is why I introduced `_is_input_map_set`, though perhaps I should try to use an `exit_tree` event instead?

If so, I should update the API for `register_custom_input_handler()` to make it clear that it accepts `null` to unset the custom input handler?

bolinfest commented:
Alternatively, should I implement `escoria-ui-keyboard-9verbs` by extending the one `game.gd` file in `escoria-ui-9verbs`? Then there would be a dependency, but it would eliminate a lot of the copied code?

dploeger commented:
The way you did this is the right one. Please don't depend on the other UI plugin, because that would break things. The UI plugins should only rely on core (for most of the time at least).

And yes, please call the super method with _enter_tree, because ESCGame needs to setup things there as well and Godot won't call it automatically.

Thanks!

bolinfest commented:
@dploeger that’s fine, though I expect the actual 9verbs will continue to evolve and improve, but it seems unlikely that this will get those fixes. To me, this really feels like an enhancement of 9verbs, which would be easier to reason about if there were only one file.

bolinfest commented:
Also given what you found about how _enter_tree() is special-cased, I guess I can leave my code as-is:

https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inheritance

dploeger commented:
> @dploeger that’s fine, though I expect the actual 9verbs will continue to evolve and improve, but it seems unlikely that this will get those fixes. To me, this really feels like an enhancement of 9verbs, which would be easier to reason about if there were only one file.

Actually, you're right, we *could* add this feature to the 9verbs instead. @StraToN @BHSDuncan wdyt?

dploeger commented:
> Also given what you found about how _enter_tree() is special-cased, I guess I can leave my code as-is:
> 
> https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inheritance

Yes! Thanks

bolinfest commented:
I should clarify: I believe I could:

(1) roll this logic into 9verbs behind a boolean to disable it by default
(2) create a new addon that depends on 9verbs, but contains only one game.gd that extends the game.gd in 9verbs (though I suppose it also needs its own .tscn file to specify in Project Settings?)

(1) is much easier to implement, but it also means that whoever starts building off of it to create their own UI will get the predefined keyboard shortcuts by default, which they may disagree with. I suppose if we make it easy enough to override the key bindings without forking 9verbs, that would be the best of both worlds?

dploeger commented:
Why not 3) make a distinct plugin without depending on 9verbs? (just copying it over like you already did)

bolinfest commented:
Again because I fear the new plug-in will diverge from 9verbs, which will make it less ideal to use as a starting point. It also makes it a lot more work to tell what is different between the original 9verbs and the one with keyboard shortcuts. If the whole plug-in is only one or two files, it’s much easier to verify the delta.

dploeger commented:
Ohh, okay. You think that the UI plugins are supposed to be a starting point to create new UI addons from?

No, that wasn't our intent. The UI plugins are provided as examples to cherry pick from and as easy, but not ideal starting points for new games.

So it's perfectly fine to have a very custom uI just to showcase this feature. It doesn't matter, if it diverges at some point.

dploeger commented:
Hm. Can you minimize the custom UI addon you added for this feature? So keeping all the 9 verbs boilerplate out of it?

bolinfest commented:
> Hm. Can you minimize the custom UI addon you added for this feature? So keeping all the 9 verbs boilerplate out of it?

Good question: I’ll give that a shot!

bolinfest commented:
OK, I updated the second commit. Here are the main files of interest:

```
addons/escoria-ui-keyboard-9verbs/game.gd
addons/escoria-ui-keyboard-9verbs/input_map.gd
```

It works, but it's a little janky. Specifically, `_process_input()` calls `escoria.action_manager.set_current_action()` to handle a keyboard shortcut. This mostly does the right thing, though it does not seem to play well with "default actions." For example, in room02, I hit `C` to enable the **Talk** action before I mouse over the button, and even though **talk** is displayed in the action bar, **use** shows up highlighted in the verb grid.

Also, although @dploeger suggested trying to slim down escoria-ui-keyboard-9verbs so it's easier to focus on the use of `register_custom_input_handler()`, I found this a bit hard to do because I think it's already pretty thin and I do want the verb grid, room selector, and inventory because it would be hard to meaningfully play/test the game without those features.

bolinfest commented:
OK, I think I fixed it! Two changes:

- replaced `escoria.action_manager.set_current_action()` with ` verbs_menu._on_action_selected()`
- dropped the code associated with `# Hovering an ESCItem highlights its default action`

bolinfest commented:
tip: when reviewing the second commit, assuming you have the code checked out locally, running the following will help identify what changed:

    diff -r addons/escoria-ui-9verbs addons/escoria-ui-keyboard-9verbs

bolinfest commented:
I just rebased on the latest version of `develop` to pick up https://github.com/godot-escoria/escoria-demo-game/pull/510, which makes this change a little easier to test.

bolinfest commented:
I updated this PR to drop the call to `get_tree().set_input_as_handled()` as part of `_unhandled_input()` because it doesn't seem like the right thing to do based on my gamepad work.
